File: Cisco/LinkedInList/Python/3MaximalSquare.py
class Solution(object):
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        NROWS, NCOLS = len(matrix), len(matrix[0])
        cache = [-1] * NROWS * NCOLS

        def dfs(r, c):
            # Out of bounds.
            if r >= NROWS or c >= NCOLS:
                return 0

            # Already cached.
            if cache[r * NCOLS + c] != -1:
                return cache[r * NCOLS + c]

            cache[r * NCOLS + c] = 0
            right = dfs(r, c + 1)
            diag = dfs(r + 1, c + 1)
            down = dfs(r + 1, c)

            if matrix[r][c] == "1":
                cache[r * NCOLS + c] = 1 + min(right, diag, down)

            return cache[r * NCOLS + c]

        dfs(0, 0)

        return max(cache) ** 2

File: Cisco/LinkedInList/C++/3MaximalSquare.cpp
    #include <vector>
    #include <algorithm>

    using namespace std;

    class Solution
    {
    public:
        int maximalSquare(vector<vector<char>> &matrix)
        {
            int M = matrix.size(), N = matrix[0].size();
            if (M == 0 || N == 0)
                return 0;

            vector<vector<int>> dp(M, vector<int>(N, 0));
            int maxSquare = 0;

            for (int i = 0; i < M; ++i)
                for (int j = 0; j < N; ++j)
                    if (matrix[i][j] == '1')
                    {
                        if (i == 0 || j == 0)
                            dp[i][j] = 1;
                        else
                            dp[i][j] = 1 + min({dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]});

                        maxSquare = max(maxSquare, dp[i][j]);
                    }

            return maxSquare * maxSquare;
        }
    };
